// @ohos/hpm-cli-ui/dist/server/service/trace-service.js
"use strict";

var path = require('path');

var fs = require('fs');

var _require = require('../utils/utils'),
    getFormattedDate = _require.getFormattedDate;

var _require2 = require('../utils/constant'),
    DEFAULT_HISTORY_LOG_DIR = _require2.DEFAULT_HISTORY_LOG_DIR;

var traceLogger = {
  historyLogPath: path.join(DEFAULT_HISTORY_LOG_DIR, "history-".concat(getFormattedDate(new Date(), 'YYYY-MM-DD'), ".log")),
  writeToLog: function writeToLog(str) {
    if (!fs.existsSync(DEFAULT_HISTORY_LOG_DIR)) {
      fs.mkdirSync(DEFAULT_HISTORY_LOG_DIR);
    }

    var writeStream = fs.createWriteStream(this.historyLogPath, {
      flags: 'a'
    });
    writeStream.write(str, 'utf-8');
  },
  write: function write(data) {
    var str = "[HPMCLI_UI] : \n".concat(JSON.stringify(data), "\n");
    this.writeToLog(str);
  },
  unlinkLog: function unlinkLog(filePath) {
    fs.unlinkSync(filePath);
  },
  getHistoryLogger: function getHistoryLogger(filePath) {
    return new Promise(function (resolve, reject) {
      var readStream = fs.createReadStream(filePath);
      var data = '';
      readStream.on('data', function (chunk) {
        data += chunk;
      }).on('error', function (error) {
        reject(error);
      }).on('close', function () {
        resolve(data);
      });
    });
  }
};
module.exports = {
  traceLogger: traceLogger
};
